### ISSUE TITLE
End of service for Bintray in May, 2021

### ISSUE THREAD

**Issue by imprakharshukla** [termux/termux-packages#6348]
**Problem description**
Jfrog's Bintray which Termux uses as the primary hosting, is ending support in May of 2021. Here is the official announcement from the VP of dev relations for Jfrog https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/.

> Bintray/JCenter users should start migrating to a new hosting solution.

What new hosting solution will Termux switch to?

**Comment by RossComputerGuy** [comment]
I've noticed Linode is a cheap solution for running a VPS, ~$5/mo but traffic may cause the price to go up.

**Comment by imprakharshukla** [comment]
@RossComputerGuy $5 isn't the actual price because Bintray had allocated around 32 TB / month, which would account to being approximately $350 in Linode's outbound transfer. This doesn't come with a CDN as opposed to Bintray which makes it terribly slow.

Using a dedicated object storage like S3, DO Spaces, wasabi or blackblaze will require an app update, therefore will render the Play Store version useless. Plus it will shoot the prices well beyond a $1000.

**Comment by WMCB-Tech** [comment]
In the Feb 28:
> No more submissions will be accepted to Bintray, JCenter, GoCenter, and ChartCenter
The GoCenter and ChartCenter websites will be disabled (client requests will still work)

So this means that package updates will be stopped? after Feb 28?

**Comment by imprakharshukla** [comment]
@WMCB-Tech That is very confusing. They could also mean that no new user would be able to register for these services or just that all the rest and client requests will be blocked. I think former. How unprofessional of them to give such a short window for the migration.

**Comment by WMCB-Tech** [comment]
Termux has more than 1000+ packages available though, including x11 ones

**Comment by kcubeterm** [comment]
By the way I got some good results with ipfs at some extent but don't know what happen if lots of user will be there. So I will share links once I managed to get atleast 10 node.

**Comment by xeffyr** [comment]
> So this means that package updates will be stopped in Feb 28?

Yes, this means any data submissions (package uploads) will be stopped.

> By the way I got some good results with ipfs at some extent but don't know what happen if lots of user will be there.

@kcubeterm I can give a "stress test" to your setup by sending traffic from https://main.termux-mirror.ml to IPFS repo for some time.

> So I will share links once I managed to get atleast 10 node.

Hopefully these 10 nodes are not 10 (free) accounts on Pinata/Temporal... not saying that this is (likely) against their ToS but also effectively only 2 nodes.

**Comment by xeffyr** [comment]
Ok, then I will manage to upload all latest available updates and then disable Github Actions uploads on next week to prevent any potential inconsistent state on Bintray repo. We still have mirrors as shown in https://github.com/termux/termux-packages/wiki/Mirrors, so hosting by itself is not a problem currently but package submission workflow likely will be switched to manual.

@a1batross Are you continue hosting Termux mirror at https://termux.mentality.rip? I have noticed that last sync was on 26 Dec 2020.

**Comment by a1batross** [comment]
@xeffyr sorry, I forgot to move mirror.list config from old server so I rewrote it from scratch. Now it's mirroring `{termux,game,science,unstable,x11}-packages`. :)

**Comment by insign** [comment]
Maybe we (community) can seed like a p2p network, but or create public mirrors based on checksum verifications?

**Comment by Harshiv-Patel** [comment]
Well if every node needs to have the whole 20-50gigs worth of data, the torrent like p2p would be quite expensive for end users, Though I'm interested in the idea. I guess it'd have to be something where people can join and leave seeding parts of it from time to time.

The whole thing looks far more complicated the more I think about it.

**Comment by a1batross** [comment]
@Harshiv-Patel termux repos are less than 10GB.

The problem is not the distribution though but publishing. Bintray was pretty good at this.

**Comment by kcubeterm** [comment]
Alright, here's IPFS repo. Based on P2P. Those who want to host few or whole repo can pin content from here. http://ipfs.io/ipns/k51qzi5uqu5dhght6oyh7c83cqfbssco7xx7yti0me15eoky0eyppd2go9scgw

**Comment by xeffyr** [comment]
As for testing, https://main.termux-mirror.ml is also on IPFS now. Can be accessed either directly (CloudFlare gateway) or with https://ipfs.io/ipns/main.termux-mirror.ml (IPFS default gateway).

**Comment by xeffyr** [comment]
> Those who want to host few or whole repo can pin content from here.

And will need to re-pin it any time the repo is changes.

**Comment by kcubeterm** [comment]
> Repository inconsistency: few deb files missing

Most probably, this issue should be fixed after fresh Installation. I have published newer cid on same ipns

> Accessibility (stability?) issues: traffic has been reduced by 70% which means users have problems with downloading packages.

Not sure about it, I have installed many packages and there's no issues, latency is also better than before.

**Comment by xeffyr** [comment]
https://main.termux-mirror.ml/ is now served over CloudFlare IPFS gateway. Alternate access way would be any other gateway with path `/ipns/main.termux-mirror.ml`. Higher latency is expected.

**Comment by xeffyr** [comment]
@kcubeterm I'm switching to custom key for IPNS which is detached from the node id. As result link has been changed now. Unstable and X11 repositories will soon be in IPFS and each will have own link.

If you decide to use custom keys instead of node-specific key pair, here is a quick tutorial:
```
# Generate new
ipfs key gen termux-main-repo

# Publish IPNS
ipfs name publish --key=termux-main-repo <CID>
```

**Comment by kcubeterm** [comment]
Alright, I am aware of that tutorials, but what can I do here, since it's your key only you could publish unless you share that key. From now I will only pin repo with your CID (since it's change always so first my node will resolve ipns) 
So ensure don't change your ipns link next time otherwise I have to manually put it into all nodes.

**Comment by xeffyr** [comment]
@a1batross Your mirror looks messed up. There were reports about checksum mismatch from some users and some repositories are empty, e.g. https://termux.mentality.rip/termux-root-packages-24/

**Comment by a1batross** [comment]
@xeffyr just did that and changed all unix permissions/ownership.

I joined #termux IRC channel, so if there will be any problems you can ping me there.

**Comment by buttaface** [comment]
Honestly, this is a good thing, as I never liked that Termux was so dependent on that company for free bandwidth. I agree with @insign, we should be able to set up a p2p network between the app installs, seeded by a single cheap VPS that provides the freshly built packages and checksums. The only question is whether we can easily repurpose some existing apt/torrent code out there for this or if it will require some custom code written by us.

**Comment by kcubeterm** [comment]
@buttaface We have tested ipfs and got good results. Right now(from yesterday) whole termux traffic are hosting from ipfs and no one reports any error regarding this. We have 12 nodes right now. And that is sufficient but since it's P2P more contributions are welcome.

**Comment by buttaface** [comment]
I'm not familar with how ipfs can be used for this: who's running those 12 nodes and are they in data centers? In that case, it's federated and not true p2p like bit torrent, but that's fine for now if it works. If you put up instructions on setting up IPFS nodes for Termux, I'm sure more people will do it.

**Comment by xeffyr** [comment]
Everyone can set up a local ipfs node and use it as local gateway. Thus sources.list will look like
```
deb http://127.0.0.1:8080/ipns/k51qzi5uqu5dg9vawh923wejqffxiu9bhqlze5f508msk0h7ylpac27fdgaskx stable main
```
and will re-provide content which has been already cached.

***

### How to set up a local IPFS node/gateway

1. Install ipfs package
   ```
   pkg install ipfs
   ```
2. Initialize the node. This will create `~/.ipfs` directory with basic data structure and configuration file.
3. Start daemon. You can run it either in a separate session, in `tmux`, etc.
   ```
   ipfs daemon
   ```
   Termux ipfs package currently lacks support of `termux-services` but this will be fixed.

Also you need to forward port `4001`, both tcp and udp, or enable relay discovery:
```
ipfs config --json Swarm.EnableRelayHop false
ipfs config --json Swarm.EnableAutoRelay true
```

Better to visit https://docs.ipfs.io/ to learn what IPFS actually is and how to work with it.

***

For now, IPFS setup is centralized as everything goes through single gateway node (https://ipfs.io). If even 1/1000 of users will use a local gateway instead, this will make Termux less dependent on third-party gateway and will reduce load on our seed nodes.

Though it is better to make a full copy of seeded content.
```
ipfs pin add /ipns/k51qzi5uqu5dg9vawh923wejqffxiu9bhqlze5f508msk0h7ylpac27fdgaskx

# Advanced variant, with garbage-collecting stale data.
ipfs pin rm /ipns/k51qzi5uqu5dg9vawh923wejqffxiu9bhqlze5f508msk0h7ylpac27fdgaskx
ipfs pin add /ipns/k51qzi5uqu5dg9vawh923wejqffxiu9bhqlze5f508msk0h7ylpac27fdgaskx
ipfs repo gc
```
and repeat this on periodic basis to ensure that pinned content is up to date.

**Comment by kcubeterm** [comment]
And those who is behind NAT and can't forward TCP/UDP port. Enable autorelay 
https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#autorelay

**Comment by xeffyr** [comment]
@kcubeterm autorelay instructions are already specified in my comment.

**Comment by kcubeterm** [comment]
@xeffyr 🤭🤭

**Comment by xeffyr** [comment]
There however is one issue with user setup: `ipfs daemon` should run continuously. If user just using Termux for 5-10 minutes and then close, it will be useless. 2 - 3 hours minimum, especially if using relays.

**Comment by xeffyr** [comment]
There are variants with using `apt-p2p` but I don't think it will be better than IPFS.

**Comment by kcubeterm** [comment]
Running ipfs daemon as user perspective is not good idea. as you mentioned it gets killed and it will eat extra bandwith as well. 
Only thing I'm worrying about is gateway. By the way that was good idea as pkg choose mirrors in same way it will choose different gateway also

**Comment by kcubeterm** [comment]
Here's list of few public gateway https://ipfs.github.io/public-gateway-checker/

**Comment by xeffyr** [comment]
> as you mentioned it gets killed

I have not mentioned that it "gets killed" exactly. But certain amount of cpu time, ram and bandwidth will be used.

P2P is useful when clients contribute too and there no way to avoid bandwidth use. Currently we pushed all traffic problems to ipfs.io which really isn't good.

> Here's list of few public gateway

I know and will update `pkg`.

Not all gateways can be used. Some gateways looks like able to filter traffic somehow.

**Comment by xeffyr** [comment]
> But who is running those 12 nodes now?

I run a one node. Rest is by @kcubeterm and few other people.

**Comment by landfillbaby** [comment]
idea: what if each PPA repo, was hosted as a branch on the respective GitHub repo?

**Comment by xeffyr** [comment]
1. Repository size limit. It's soft, but going beyond will trigger interest of Gihub staff.
2. Traffic limit of 100 GB per month. Again, soft.

Hosting binary files on Github can lead to terms of service violation.

Opengapps project had problem with Github due to these limits, https://opengapps.org/blog/post/2019/02/17/github-situation/.

**Comment by landfillbaby** [comment]
ah. that's a shame

**Comment by Luflosi** [comment]
I propose creating a collaborative IPFS cluster, similar to https://github.com/RubenKelevra/pacman.store. This would make it much easier for people to pin everything on their IPFS node without the frequent `ipfs pin rm` and `ipfs pin add` suggested in https://github.com/termux/termux-packages/issues/6348#issuecomment-778279980. It would also increase the visibility of this effort if it is listed on https://collab.ipfscluster.io.

**Comment by SDRausty** [comment]
`free unlimited bandwidth web hosting debian`

**Comment by AndreiJirohHaliliDev2006** [comment]
Will package installs after May 1 on legacy environments break?

**Comment by Grimler91** [comment]
@AndreiJirohHaliliDev2006 The main legacy repository is not hosted on bintray so will not be affected. The science and game-packages repos are though. I will set up a mirror for them elsewhere, but users might need to manually adjust the URL to be able to use them

**Comment by m-p-3** [comment]
> ```
> ipfs pin add /ipns/k51qzi5uqu5dg9vawh923wejqffxiu9bhqlze5f508msk0h7ylpac27fdgaskx
> 
> # Advanced variant, with garbage-collecting stale data.
> ipfs pin rm /ipns/k51qzi5uqu5dg9vawh923wejqffxiu9bhqlze5f508msk0h7ylpac27fdgaskx
> ipfs pin add /ipns/k51qzi5uqu5dg9vawh923wejqffxiu9bhqlze5f508msk0h7ylpac27fdgaskx
> ipfs repo gc
> ```
> 
> and repeat this on periodic basis to ensure that pinned content is up to date.

Okay, so you gotta make a cron job, as IPFS doesn't have any mechanism in place to monitor and keep an /ipns/ path synced for now? And there's no way to do it without purging the entire data and redownloading the whole repo each time, if I understood correctly?

**Comment by Luflosi** [comment]
If you do it this way, IPFS will only download new data but there will still be a lot of disk activity as IPFS will go through its tree of hashes and check if all the blocks are on disk.

**Comment by m-p-3** [comment]
Got it, and thanks for letting me know about IPFS-cluster, that looks like a great system that would gracefully solve this problem.

**Comment by AndreiJirohHaliliDev2006** [comment]
> @AndreiJirohHaliliDev2006 The main legacy repository is not hosted on bintray so will not be affected. The science and game-packages repos are though. I will set up a mirror for them elsewhere, but users might need to manually adjust the URL to be able to use them

@Grimler91I asked about that in case the devs had any plans to permanently shut down the Android 5.x-6.x packages repo (thus break `pkg install`s for Play Store users, and also backporting to these unsupported versions is also a challenge), but since this stuff is on IPFS, I don't need to worry about it. I would like to maintain these packages to keep them up-to-date, as much as I can.

Sidenote: Since I'm new to deb packaging in general, this might give me some time to practice the packaging process.

**Comment by Grimler91** [comment]
> @Grimler91 asked about that in case the devs had any plans to permanently shut down the Android 5.x-6.x packages repo (thus break `pkg install`s for Play Store users, and also backporting to these unsupported versions is also a challenge), but since this stuff is on IPFS, I don't need to worry about it. 

The main legacy repository is not on IPFS (as far as I know), but there are archives with the packages saved [here](https://archive.org/details/termux-repositories-legacy). The main legacy repo is hosted on termux.net and controlled by @fornwall. I don't know if he has plans on continuing hosting it indefinitely or if he might sooner or later need the disk space for other things.

**Comment by iBug** [comment]
Thank you all for this amazing project for Android! Please allow me to share my $0.02 as a long-term Termux user.

As far as I know, Backblaze B2 provides affordable object storage for $5 per TB-month, and free outbound traffic if you place Cloudflare CDN in front of your buckets. This may be a good option for a small project, but with over 30 TB of monthly traffic, Cloudflare may complain about violating its [Terms of Service][1], Section 2.8, which specifically calls out that excessive non-web content is prohibited, *unless purchased separately as part of a Paid Service*.

I would recommend giving it a try to contact Cloudflare to see if they can offer a good plan to deliver Termux APT/dpkg packages through their network, thus eliminating the traffic cost from Backblaze B2. In theory, if this pulls off, I would expect a monthly cost of less than US$50, which I sincerely hope is an affordable cost for Termux maintainers.

**Disclaimer:** I an NOT affiliated with either Backblaze, Inc. or Cloudflare, Inc.

---

Besides that, I'm also interested in any possible way public mirrors can assist Termux in delivering content. In my residual country (China), I am aware of multiple (6+) university-held public software mirror sites, including but not limited to those in [Tsinghua University][2] and [University of Science of Technology of China][3]\*. Those mirrors are already serving TBs of packages for Termux, and their maintainers are eager to provide support should an open-source community need assistance.

If it's possible for mirror sites to provide additional help, please create a public thread and I can loop them in.

<sup>\* I am one of the maintainers of the said mirror site.</sup>

  [1]: https://www.cloudflare.com/terms/
  [2]: https://mirrors.tuna.tsinghua.edu.cn/termux/
  [3]: https://mirrors.ustc.edu.cn/termux/

**Comment by fphammerle** [comment]
@iBug, you could setup `ipfs`, pin the hashes posted in https://github.com/termux/termux-packages/issues/6348#issuecomment-776169136 and optionally serve a public ipfs gateway in your area. that's basically equivalent to hosting a repo mirror

**Comment by vladimyr** [comment]
Homebrew is also using Bintray and also in process of finding a new home for their artifacts. According to  https://github.com/Homebrew/discussions/discussions/691#discussioncomment-469668 they are going with Internet Archive which seems like a very interesting possibility. Might be a good fit for Termux too?

**Comment by xeffyr** [comment]
@vladimyr Isn't this is a their terms of service violation? As far I know, it doesn't like to be a CDN for content delivery and there already were incidents.

Also, their TOS states:
> You agree to indemnify and hold harmless the Internet Archive and its parents, subsidiaries, affiliates, agents, officers, directors, and employees from and against any and all liability, loss, claims, damages, costs, and/or actions (including attorneys’ fees) arising from your use of the Archive’s services, the site, or the Collections.

Don't forget that Termux generates lots of traffic (40+ TB, even more if we submit bunch of updates) and this is major issue.

However we can try it, considering it provides S3-like API, https://archive.org/services/docs/api/ias3.html.

**Comment by buttaface** [comment]
Btw, any reason the automated package updates have not been turned on again, 468b4d4? I thought turning it off was a temporary move while transitioning to new distribution, but surely we can have the updater upload to some new seed location now that we're not using Bintray?

**Comment by xeffyr** [comment]
As far I know, Termux doesn't have seed location now.

**Comment by buttaface** [comment]
Meaning what, when a package is updated and rebuilt now, someone has to manually give those new packages to the ipfs mirrors?

**Comment by vladimyr** [comment]
@xeffyr Honestly, I don't have the answers you asked for and fully understand your concerns. This came as surprise to me too because I'd also guess they wouldn't let you do that.

Knowing Homebrew folks I'm sure they aren't trying their luck and instead they probably reached out to Internet Archive folks and made a deal. Especially because they generate even bigger bandwidth https://github.com/Homebrew/discussions/discussions/691#discussioncomment-340426

Let's try something, instead of me passing speculations around I'm cc-ing HB's lead maintainer @MikeMcQuaid Hopefully, Mike has a minute (he is a really busy man) and is willing/allowed to share HB's experiences with you to help a sister-like project or at least delegate discussion to someone else from a plethora of awesome HB maintainers. Let's try to start a conversation and build some bridges for mutual benefit...

Disclaimer, I'm not entitled to any outcome, understand if there is no interest from the HB side, and if such discussion can't be held publicly and for any possible reason won't happen in private either. I'm operating here solely on goodwill as both user and contributor (not much on Termux end tbh) to both projects which I love and would do anything I could to help them grow and sustain. Take this as the virtual bow of appreciation of both sides and all the great work they are built upon!

**Comment by kcubeterm** [comment]
> someone has to manually give those new packages to the ipfs mirrors?

Yea, @xeffyr do that.

**Comment by buttaface** [comment]
Instead of that manual process, couldn't we designate one of the ipfs distributors as the seed, automate uploading to it from the CI, and have the others mirror it, similar to how Bintray and the web mirrors worked before? I'm not familiar with ipfs, but I imagine we could set up something like that.

**Comment by Luflosi** [comment]
See my comment https://github.com/termux/termux-packages/issues/6348#issuecomment-779780505 from above for some ideas regarding IPFS.

**Comment by MikeMcQuaid** [comment]
> Knowing Homebrew folks I'm sure they aren't trying their luck and instead they probably reached out to Internet Archive folks and made a deal. Especially because they generate even bigger bandwidth [Homebrew/discussions#691 (reply in thread)](https://github.com/Homebrew/discussions/discussions/691#discussioncomment-340426)

We're no longer planning on using The Internet Archive but instead a GitHub solution specific for Homebrew (for now).

**Comment by mdmisch** [comment]
If I understood it right, one of the current bottlenecks with IPFS for repository hosting is that we need gateways so that APT could download files. Maybe, it would be better to use IPFS transport instead?  Like this one: [apt-transport-ipfs](https://github.com/JaquerEspeis/apt-transport-ipfs)

**Comment by xeffyr** [comment]
> Maybe, it would be better to use IPFS transport instead?

On-gateway data caching is the key why I have chosen IPFS, lots of free bandwidth. A far better than previous setup with CloudFlare CDN. `apt-transport-ipfs` seems discards the all purpose. Also, this transport method seems require `python` and `ipfs` as dependencies which would be too fat to fit into application bootstrap archive.

**Comment by iBug** [comment]
What about using GitHub Container Registry as the Homebrew guys [are doing][1]? They have some reference [code][2] for uploading non-Docker-images to GHCR. Free storage and bandwidth for open-source projects, sounds good.

  [1]: https://github.com/Homebrew/discussions/discussions/691#discussioncomment-584299
  [2]: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/github_packages.rb

**Comment by Grimler91** [comment]
@iBug Uploading just .debs there would probably work, but apt expects a certain folder structure and (signed) Release and Packages files, and I am not sure if that is possible there.

> This is also more friendly for mirror sites that are already using Debian APT-based sync tools.

Do you have an example of where these files are synced to another place? How the files are hosted at github packages does not seem very sync-friendly to me

**Comment by iBug** [comment]
@Grimler91 Yes, Homebrew guys' script allows arbitrary directory structure as long as you can stand some prefixes.

An example of APT-based sync tool can be found at [ustclug/ustcmirror-images](https://github.com/ustclug/ustcmirror-images/blob/master/apt-sync/tunasync/apt-sync.py), which fetches and parses the Releases file for each configured "dists" and download the deb packages to the right place. This script, along with other similar scripts created and used by other parties, is completely compatible with a standard APT repository, requiring only HTTP access to the source storage.

**Comment by xeffyr** [comment]
> which fetches and parses the Releases file for each configured "dists" and download the deb packages to the right place.

Is it able to upload individual deb files, update the metadata and publish to Github Packages without re-downloading whole repository or even worse, depending on existing mirror?

**Comment by iBug** [comment]
@xeffyr Unfortunately that doesn't seem quite easy given the structure of an APT repository, if at all possible.

**Comment by iBug** [comment]
I just came up with another hosting scheme:

- We host a server for the `Release` and `Release.gpg` files, and possibly along with the `dists/` directory
- The actual packages along with the signatures are packed into Docker images and uploaded to GHCR.
  - This can be done by-project, so we can package `curl` and `libcurl4` into one single container, reducing push size for each update
- The server we're hosting **issues 302 redirects** for the `pool/` directory, directing clients to GHCR to download the actual deb packages

Costs should be relatively low this way as the biggest bandwidth consumer may only be the `Packages.gz` file in each "dist".

**Comment by iBug** [comment]
Sorry for bothering. Please ignore the idea of leveraging GitHub Packages.

After careful investigation, I discovered that Homebrew bottles are all `.tar.gz` that can be squeezed into one Docker image layer (OCI layer), which can be fetched independently of other layers (blobs). So they basically just crafted the bottles into *container layers* and uploaded them to a container registry as a supported format for both the `brew` client and the registry (storage provider)

This format is not suitable for Debian packages, unfortunately. I'm afraid I don't have better ideas than IPFS at present.

**Comment by Grimler91** [comment]
@iBug Thanks for looking into it

**Comment by dosy4ev** [comment]
@iBug Is it possible to modify your idea for github releases?

If there will be github repo for each termux package (group of packages) you can host debs and signatures as release artifacts instead of docker layers.

**Comment by iBug** [comment]
@dosy4ev I'm pretty sure it's possible as long as we look into it deep enough.

However, do keep in mind that release assets have to reside at a URL of this form:

    https://github.com/{owner}/{repo}/releases/download/{tag}/{filename}

And the good thing is, there's a [GitHub Releases API][1] for listing Releases and uploading / modifying / deleting Release Assets, which means we could build some tools to utilize that. We could upload the Debian `Release` and `Release.gpg` files to the repository itself and host them on GitHub Pages, while loading the `.deb` packages over GitHub Releases.

Makes total sense to me.

  [1]: https://docs.github.com/en/rest/reference/repos#releases

**Comment by Harshiv-Patel** [comment]
Is there a tutorial somewhere to join in on the IPFS thing?
What would be the network requirements for that, is a 24x7 broadband like connection mandatory? Or can I join in with limited mobile data & bandwidth, and stop it as and when I need to.
Is the impact on system performance similar to running torrents?

**Comment by xeffyr** [comment]
> Is there a tutorial somewhere to join in on the IPFS thing?

All tutorials here: https://docs.ipfs.io/how-to/

Just periodically pinning IPNS hash would be enough, if you don't want to provide a public gateway or run a custom `apt` repository copy.

> What would be the network requirements for that, is a 24x7 broadband like connection mandatory?

It is not necessarily to run 24x7. I'm running it only 15 hours per day, but due to caching of the content on gateway, for most users it looks like "still running". Rarely used packages may not be cached though, which results in whatever errors during download.

> Or can I join in with limited mobile data & bandwidth

Expect 3 - 8 GB of bandwidth per day.

> Is the impact on system performance similar to running torrents?

Bandwidth use lower than for torrents. However CPU usage is relatively high, so don't run it on mobile devices as battery charge will gone quickly.

***

> We could upload the Debian Release and Release.gpg files to the repository itself and host them on GitHub Pages, while loading the .deb packages over GitHub Releases.

@iBug Isn't Github Pages and release assets would be on different domains? This case seems not supported by `apt` because `Packages.gz` specifies only file path of deb file and not the URL.

In overall, this solution would not be better than IPFS due to custom `apt` layout and metadata updating enforced to incremental mode (otherwise we will need to download full repository to rebuild `Packages.gz`).

**Comment by a1batross** [comment]
@xeffyr if you let me, I have different experience with IPFS. It seems running service even 24/7 isn't enough because there is a high chance that IPFS will never find a "seeder".

That's what happened with my mirror twice, when I setup mirroring from http gateway and when I setup IPFS daemon. I switched it back to mirroring from grimler.se.

**Comment by Harshiv-Patel** [comment]
Is there an option to download the whole repo(hopefully 6-8GB), during light trafficking hours, or via torrent as an initial setup step, and seed it at my convenience, and update the repo contents once a day or once a week with just delta of new updates as they become available ?

**Comment by Luflosi** [comment]
Just use IPFS, it takes care of only downloading what you don't already have. There are no torrents AFAIK.

**Comment by xeffyr** [comment]
> It seems running service even 24/7 isn't enough because there is a high chance that IPFS will never find a "seeder".

Ensure that node ports are accessible from the Internet. You can also increase limits of maintained connections with other nodes, here is what I use in config:
```
    "ConnMgr": {
      "Type": "basic",
      "LowWater": 2000,
      "HighWater": 6000,
      "GracePeriod": "20s"
    }
```

***

> Is there an option to download the whole repo(hopefully 6-8GB)

Pinning will always download the full repo if you haven't done this before.

To pin repository, run
```
ipfs pin add -r /ipns/k51qzi5uqu5dg9vawh923wejqffxiu9bhqlze5f508msk0h7ylpac27fdgaskx
```
and then run this command on periodic basis. If you are limited on disk space, then you will need to remove pins made previously before creating a new one.
* `ipfs pin ls --type=recursive` - to list available pins
* `ipfs pin rm -r <hash>` - to delete specific pin
Once unneeded pins are removed, you can perform garbage collection `ipfs repo gc`.

*I didn't tried "ipfscluster" solution mentioned here previously, so providing instructions only for basic pin management.*

**Comment by dumb-stuff** [comment]
Why don't you guys use [Heroku](https://heroku.com) or [Replit](https://replit.com)  or [Netlify](https://netlify.com)

**Comment by xeffyr** [comment]
> Heroku

This is application hosting platform. We need either file/package hosting or at least web server with preferably unlimited data.

> Replit

This is not a file (package) hosting platform and machine stats also unsuitable.

> Netlify

Unsuitable. Maximal plan is 600 GB per months. Too low and of course non-free.

We are serving about 40 terabytes of data per month, with possible spikes to >50 TB in case of many updates submitted.

***

*I also should note that free solutions are preferred and currently nothing can beat IPFS in that.*

**Comment by buttaface** [comment]
Yes, but other than the ipfs mirrors we run, we are pushing this large traffic load onto other ipfs donors.

What about doing some fundraising so the users pay for their traffic? Set up some kickstarter-type donation monthly goal (many fundraising sites support this) and stick a link for it in the pkg installer text, so that people see the link whenever they download packages. Whatever money comes in could be used to pay for some file-serving CDN, with ipfs only used to cover unpaid bandwidth beyond that.

Most users won't donate but if 1% do, it would likely cover the monthly bandwidth.

**Comment by xeffyr** [comment]
Termux already collects donations via https://www.patreon.com/termux, PayPal and Google Play paid add-ons. However as all of them are sent to @fornwall, I'd expect that paid server would be set up by him. According to Termux [Patreon](https://www.patreon.com/termux) page, there is donated about 100$ per month which should be enough.

I have already raised the question about Bintray sunset and whether we can re-use https://termux.net or get a new one and here is answer:

> we could definitely use https://termux.net for hosting packages
it's behind cloudflare, which is great as it uses their CDN to get good performance across the globe
the problem is that cloudflare doesn't specify a hard limit for traffic costs, and they might require an enterprise plan (which is big $$$) if we start using too much traffic
and they haven't been responsive when asking for sponsorship
so one alternative would be to set May 1st as a deadline for transitioning to packages-in-apk:s / android-10

Yet nothing went beyond the discussion, so current choice is usage of existing mirrors to keep repositories online.

**Comment by eljoest** [comment]
I've noticed that [Fastly](https://www.fastly.com/open-source) seems to support open source projects with free CDN services. They provide this for the debian repositories.

Have they been contacted? I feel that I can't answer this question in their contact form:
> If you are accepted to the program, we will require you to sign a 12mo contract. Would you be the person signing that?

**Comment by dannysauer** [comment]
I went to install curl today so I could test the bintray replacent solution I set up for my employer in the last couple of months. Failure because bintray went away as promised on Saturday.

So, I go to the FAQ, and it says to run termux-change-repo or something like that which doesn't exist on my system. It also says "we don't know why it doesn't work."

Ideally there would be a solution chosen by by now, but can someone at least update the public docs to indicate this is a known failure and provide a workaround? List of mirrors, program that actually exists, or *something*? :)

**Comment by agnostic-apollo** [comment]
> So, I go to the FAQ, and it says to run termux-change-repo or something like that which doesn't exist on my system. It also says "we don't know why it doesn't work."

You are likely referring to [`read (5: I/O error)`](https://wiki.termux.com/wiki/FAQ#Why_am_I_getting_Error_reading_from_server_when_installing_package) which is different from bintray shutdown ` repository is under maintenance or down` errors.

> Ideally there would be a solution chosen by by now, but can someone at least update the public docs to indicate this is a known failure and provide a workaround? List of mirrors, program that actually exists, or _something_? :)

The fix is mentioned/linked on [termux-app home](https://github.com/termux/termux-app), [termux-packages home](https://github.com/termux/termux-packages), [termux-packages issues](https://github.com/termux/termux-packages/issues/6726), [termux-packages wiki](https://github.com/termux/termux-packages/wiki/Package-Management), [termux-packages bug_report issue template](https://github.com/termux/termux-packages/blob/master/.github/ISSUE_TEMPLATE/bug_report.md) and [r/termux](https://www.reddit.com/r/termux/comments/n3vkzw/info_bintray_repositories_are_now_shut_down/). Should be added by someone in termux wiki site FAQ if it helps. Also in playstore app description, but none of the "active" maintainers have access to it. That's pretty much what we can reasonably do.

And if you don't have `termux-change-repo` for some reason, manually edit sources list as mentioned in #6726

**Comment by dannysauer** [comment]
> You are likely referring to read (5: I/O error) 
Maybe. This is in the section entitled "Why am I getting Error reading from server when installing package" on https://wiki.termux.com/wiki/FAQ. That looked like the most likely Doc page on a Google search, and I didn't find any of those others searching for "termux bintray". Maybe I just have a bad Google search history giving me weird results, since it didn't give any of those pages. :)

Thanks for the pointers. I really do feel your pain, as I've been migrating various customers off of bintray for the last couple of months, and some are still surprised. :D

**Comment by agnostic-apollo** [comment]
> This is in the section entitled "Why am I getting Error reading from server when installing package"

Well yeah, but the error is also mentioned. You can get lots of different errors when installing packages, but yes, faq should be updated.

>Maybe I just have a bad Google search history giving me weird results, since it didn't give any of those pages. :)

Google search often doesn't work unless you use the right keywords and github issue indexing is just terrible. Best is to search directly on github or check repos.

> Thanks for the pointers. I really do feel your pain, as I've been migrating various customers off of bintray for the last couple of months, and some are still surprised. :D

You are welcome. Yeah, must be a mess. But bintray did support termux for free for years so we should be thankful for that too.

**Comment by kcubeterm** [comment]
@dumb-stuff I find an usecase of Heroku though. I could serve a reverse proxy with caching servers, and in backend there will be any Termux mirror. And Cloudflare will be in front of Heroku app. Due to ephermal filesystem Heroku app can't cache more than 24h. 

Usrer ---> Cloudflare cdn --> reverse proxy with caching --> termux mirror

For demo purpose, i used grimler mirror. since reverse proxy caching is there so if any package requested once from the mirror, it will never be requested again  for next 24h since it will either serve from Cloudflare or by reverse proxy. 

Here's, main repo, http://dl.kcubeterm.me/termux-packages-24

**Comment by iBug** [comment]
> you could setup `ipfs`, pin the hashes posted in [#6348 (comment)](https://github.com/termux/termux-packages/issues/6348#issuecomment-776169136) and optionally serve a public ipfs gateway in your area. that's basically equivalent to hosting a repo mirror

@fphammerle Unfortunately for administrative reasons we cannot setup IPFS on our mirror sites. We can only provide HTTP(S) and Rsync access to clients and we only sync from upstream in a non-P2P manner, as is with most other repositories.

What about routing users to mirror sites *by default*, as CentOS `yum` does? This should drastically reduce load on the "official (authoritative) origin".

---

BTW, can you put **Mirrors by [LUG @ University of Science and Technology of China](https://lug.ustc.edu.cn/)** onto [Mirror List](https://github.com/termux/termux-packages/wiki/Mirrors)?

<details>
<summary>
Here's our information (use the edit button to retrieve Markdown source)
</summary>
Mirror for Chinese users for better ping and download speed.

|Repository|sources.list entry                                               |
|:---------|:----------------------------------------------------------------|
|[Main](https://github.com/termux/termux-packages)      |`deb https://mirrors.ustc.edu.cn/termux/termux-packages-24/ stable main`|
|[Games](https://github.com/termux/game-packages)     |`deb https://mirrors.ustc.edu.cn/termux/game-packages-24/ games stable` |
|[Root](https://github.com/termux/termux-root-packages)      |`deb https://mirrors.ustc.edu.cn/termux/termux-root-packages-24/ root stable`|
|[Science](https://github.com/termux/science-packages)   |`deb https://mirrors.ustc.edu.cn/termux/science-packages-24/ science stable`|
|[Unstable](https://github.com/termux/unstable-packages)  |`deb https://mirrors.ustc.edu.cn/termux/unstable-packages/ unstable main`|
|[X11](https://github.com/termux/x11-packages)       |`deb https://mirrors.ustc.edu.cn/termux/x11-packages/ x11 main`|
|Android 6.0 and below|`deb https://mirrors.ustc.edu.cn/termux/termux/ stable main`|
</details>

**Comment by dumb-stuff** [comment]
How about [Google Cloud](https://cloud.google.com/products/storage)

**Comment by dumb-stuff** [comment]
Also there's [Blue Host](https://bluehost.com/),[InMontion Hosting](https://inmotion-hosting.evyy.net/),[Dream Host](https://www.dreamhost.com/) and [A2 hosting](https://www.a2hosting.com/) also give you unlimited space and bandwidth if you want to buy that package

**Comment by kcubeterm** [comment]
@dumb-stuff  What we need right now is services which can host debian repository like bintary. just upload debfile and forget about backened, that is what bintary used to do for us.  Where one don't need to manually add packages, publish and maintain it on each and every upload. 

But in your suggested services, maintainer have to do all the stuff.

**Comment by iBug** [comment]
> What we need right now is services which can host debian repository like bintary. just upload debfile and forget about backened

Launchpad PPA is the only one I could think of. You could contact Canonical Ltd. to see if they'd like to sponsor Termux in this way.

**Comment by agnostic-apollo** [comment]
>Note: We do not allow uploading pre-built binary packages.

https://help.launchpad.net/Packaging/PPA

**Comment by xeffyr** [comment]
> How about Google Cloud

The most expensive solution of listed here. As I have already mentioned, we serve terabytes of traffic.

https://cloud.google.com/vpc/network-pricing

> Also there's Blue Host,InMontion Hosting,Dream Host and A2 hosting

Blue Host provides has a max plan of 3 TB bandwidth and 60$ per month.

https://inmotion-hosting.evyy.net is somehow blocked by my adblock/anti-tracking setup. Would check it later.

DreamHost has a VPS with unmetered bandwidth for only 13.75$ per month. Seems to be acceptable, but as I said, paid setups should be discussed with @fornwall.

A2 Hosting - everything I have found, 4 TB bandwidth on max VPS for 11.99$. 4 TB is too low. Previous variant is better.

> What we need right now is services which can host debian repository like bintary.

Hosting not necessarily should be a SaaS platform. This can be a hardware or virtual server, file hosting, package hosting, etc where we can have a bidirectional file upload and preferably with unlimited bandwidth. We need a primary server on which packages will be uploaded and then distributed over mirrors. 

For now we don't have a one. IPFS setup has not been provided as permanent solution. How much nodes would not be configured, setup will be centralized anyway because I still own IPNS keys used to generate URLs and without my node running links will gone in 12h.

**Comment by a1batross** [comment]
>This can be a hardware or virtual server, file hosting, package hosting,
etc where we can have a bidirectional file upload and preferably with
unlimited bandwidth.

By the way, BuyVM.net offers unlimited bandwidth and cheap VPS/Storage at
low prices.

**Comment by buttaface** [comment]
> Yet nothing went beyond the discussion, so current choice is usage of existing mirrors to keep repositories online.

@xeffyr, it is implied here that we need @fornwall's permission or Patreon money to change the current situation, but if the guy is not responding to emails and you're doing what you want anyway, I don't see why he has to be involved. There is nothing stopping you from raising money by adding a donation link to the `pkg` installer script, which a lot more people see than anyone ever heard of the Patreon, then using that money to buy server bandwidth yourself.

[DO and several other companies offer plans with something like 4TB for $20/month](https://www.digitalocean.com/pricing/), and it wouldn't be hard to mirror the packages onto 13-15 VPS plans like that and set up our own rudimentary CDN (I would talk to them first and see if they would work with us to come up with an approved plan with extra bandwidth), _if_ no established CDN is willing to do so for around the same price, ie around $250-300/month.

You could easily raise that much money from the Termux userbase, then you wouldn't have to depend on the charity of ipfs mirrors or whoever is sponsoring the bandwidth now, who might suddenly decide not to pay for our bandwidth use tomorrow just as bintray did.

**Comment by dumb-stuff** [comment]
Well I founded this
https://www.neuprime.com/l_vds4.php
Max up :
CPU : you can customize it
MEMORY : you can choose it
SSD : you can choose
OS : CentOS, Ubuntu, OpenSUSE, Debian, and any version of Windows Server from 2003 to 2019.
Bandwidth : unlimited

**Comment by iBug** [comment]
May I ask why this issue is closed? It doesn't seem to have resolved yet, or are we on IPFS already and is the current solution mature?

**Comment by xeffyr** [comment]
Seems like everyone ok with current solution, i.e. `grimler.se`, IPFS and optionally other available mirrors.

![gitter](https://user-images.githubusercontent.com/25881154/117577828-1a055d00-b0f4-11eb-8d32-efe85a029f83.png)

I'm not going to buy a VPS on my own.

**Comment by buttaface** [comment]
You wouldn't be buying it on your own, some of the Termux users would be donating for everyone else's bandwidth. If you're worried about fluctuating donations and getting stuck with the VPS bills, you could fundraise for three months at a time, so you have extra money up front before getting the monthly bills. I think it is irresponsible for this project to keep looking for external donors, like bintray or whoever is providing the ipfs bandwidth now, when the users could easily cover $300/month themselves.

**Comment by Grimler91** [comment]
Fundraising every three months sounds like a lot of work. Better to use the patreon money in that case. Either that Fornwall sets up vps:es and hand out access, or that he channels the funds to some one else of us that can do this.

Currently about 100 usd is donated each month through patreon. 300 usd per month through fundraising sounds like quite a lot. I would expect it to work for some months before people decrease the donation amounts. I'm also guessing it is more likely that people see themselves setting up an ipfs node on their server rather than donating money monthly.

**Comment by iBug** [comment]
Where are the unstable / x11 / game / science / root etc. packages going, then?

**Comment by buttaface** [comment]
> Fundraising every three months sounds like a lot of work

It is super easy, [there are literally dozens of sites that do all the work for you](https://www.wholewhale.com/tips/13-fundraising-websites-meet-every-campaign-need/). As for setting up a bunch of VPS mirrors with that money, I have done it before: it's just some initial setup and then no work whatsoever after that.

> Better to use the patreon money in that case

As you say in your next paragraph, it isn't enough.

> Either that Fornwall sets up vps:es and hand out access, or that he channels the funds to some one else of us that can do this

The guy doesn't even answer emails and doesn't seem to know we've moved to ipfs, I don't see it happening.

> I'm also guessing it is more likely that people see themselves setting up an ipfs node on their server rather than donating money monthly.

How many people have ever set up an ipfs node versus donated monthly, the ratio must be what, one to a million? No, it is not more likely. If you stuck the donation link in the package installer where every user would see it, I bet you'd easily hit $300/month.

**Comment by xeffyr** [comment]
> The guy doesn't even answer emails and

https://gitter.im/termux/dev

> doesn't seem to know we've moved to ipfs

He knows about that.

**Comment by Grimler91** [comment]
> Will you add 302 redirects for those repositories under `termux.org`? That'd be very convenient for downstream (users / mirror sites).

Would be convenient, but it is handled through cloudflare right now and they only offer 3 redirects for free accounts. We are currently using two (termux.org->termux.com and termux.org/packages->xeffyrs ipfs node)

**Comment by iBug** [comment]
> Would be convenient, but it is handled through cloudflare right now and they only offer 3 redirects for free accounts. We are currently using two (termux.org->termux.com and termux.org/packages->xeffyrs ipfs node)

@Grimler91 You can host that domain at Netlify to utilize their free [redirects](https://docs.netlify.com/routing/redirects/) feature. It's not limited to 3 redirects (Page Rules) and you can still put Cloudflare in front of it for the best performance.

**Comment by fphammerle** [comment]
What are the disadvantages of the ipfs-based solution? Are there any unresolved issues?
I wonder why you guys keep suggesting alternatives.
For me https://k51qzi5uqu5dg9vawh923wejqffxiu9bhqlze5f508msk0h7ylpac27fdgaskx.ipns.dweb.link/ worked great so far and re-pinning can easily be automated.

**Comment by fornwall** [comment]
@buttaface (and others): I'm indeed sorry for my limited involvement and unresponsiveness in Termux the latest year :(.

1. The email volume I get about Termux is literally hundred of mails per day. It's just hard to keep up with that volume, so for now it's best to ping me at https://gitter.im/termux/dev (I'll also make sure to follow this thread now).

2. When it comes to handing out control and assisting in setting up servers or other resources - let me know and I'll be glad to help out! Unfortunately, until vacation in June I have limited time to spend on it, but I'll happily assist in whatever way I can!

**Comment by landfillbaby** [comment]
the only downside i've noticed with ipfs from the user side is that apt always registers it as "get" (redownload) rather than "hit" (unchanged so no download), presumably due to the http date-modified header

**Comment by xeffyr** [comment]
> which seems to contradict their own specs?

I don't use their "UnixFS". Just mapping a local apt repository made with `aptly` into IPFS using [FileStore](https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#ipfs-filestore).

**Comment by buttaface** [comment]
@xeffyr, do we have an idea of who is providing the ipfs bandwidth for the Termux packages now? My understanding is that you and a few others run ipfs nodes that provide 10-15 TBs/month on servers that you pay for and the remaining 40 TBs/month is provided by the decentralized network, similar to bit torrent, so we don't know who is donating the majority of the bandwidth? @fphammerle, I'm against such reliance on external donors, who have nothing to do with Termux (like bintray), and thus may walk away at any time.

@fornwall, no need to be sorry about that, I understand you have a family and other responsibilities. I just wish you had delegated better to the Termux team and we could have moved forward on some of this stuff even if you were largely unavailable. What do you think about my idea of using one of those fundraising sites I linked and putting prominent donation links in the Termux app? That way we could fundraise donations from the users themselves and the people using the bandwidth to download packages could pay for it themselves, at least a small minority of them.

**Comment by xeffyr** [comment]
I provide no more than 15 - 40 GB each month, which is a 20 times less than was with CloudFlare. Don't know current traffic amount on nodes by @kcubeterm but should be similar.

Gateways have 99.9% caching rate, unlike CloudFlare or Fastly, but usually cache data for only 12 hours. Some gateways cache data for much longer period. Most traffic is served by IPFS gateways. Some (`ipfs.io` and probably `dweb.link`) are driven by IPFS project owners, some belong to related projects and others are hosted by unknown parties.

It is not bittorrent, it is a distributed CDN. Seed nodes use traffic only to provide data to clients, which could be either other users running the node or gateways providing data to clearnet.

In Termux repository case, P2P network is used only to deliver data on gateways which cache it, so trying to scale it with adding a lot of "seeders" is pointless. To reduce impact on gateways, users should use a localhost gateway and thus setup IPFS daemon, potentially at cost of much increased battery drain (high cpu use) and traffic (you will serve data for others).

**Comment by Luflosi** [comment]
Note that you can reduce the bandwidth and CPU usage of the IPFS daemon by a lot by setting `Routing.Type` to `dhtclient`, see https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#routingtype.
Running an IPFS daemon and changing this setting is still better than using the gateways since then the files can be transferred from any IPFS note that has the files (that did not set this setting to `dhtclient`) instead of just one gateway.

**Comment by SDRausty** [comment]
> is literally hundred of mails per day.

Hopefully nudging into thousands soon; Can you delegate, [divide and conquer](https://github.com/BuildAPKs/buildAPKs.entertainment/tree/master/madisp/divide-and-conquer/madisp-divide-and-conquer-60678a0), and share this difficult and time consuming task in some way?

**Comment by xeffyr** [comment]
> 1 even 1/1000 of users will use a local gateway

It is not easy to tell everyone what and why they should configure. Especially if everyone will be warned that IPFS daemon eats battery charge and traffic.

I can make a script and add it to `$PREFIX/etc/profile.d` supplied with `apt` package. Script will ask user for confirmation on application start to automatically configure and run IPFS node once done. But I'm very sure there would be lots of complaints about "unknown thing" running in background and eating resources. Just understand that 5 minutes of running IPFS daemon is not enough. It should run for hours and bootstrapping would take significant time.

**Comment by SDRausty** [comment]
> Can a simple selfconfiguring termux*.deb be packaged for users to use that fulfills both these wants:

> I can make a script and add it 

This sounds like a potential resolution; thank you!

**Comment by SDRausty** [comment]
> It is not easy to tell everyone what and why they should configure. 

Can `getprop [question] | grep [answer]` be employed to assit the user with these, at times, very difficult questions when Android might already know the answer?

**Comment by SDRausty** [comment]
> will ask user for confirmation on application start to automatically configure and run IPFS node once done

Obviously the user should be aware of their actions; thank you for pointing this out.

> "unknown thing" running in background and eating resources. 

Obviously the user should be aware of their actions here also; Some users might be interested in knowing more about the "unknown thing" running and the importance of the `termux-wake-lock` command.

**Comment by xeffyr** [comment]
@SDRausty I don't know whether you are understanding me or not, but issue here is *to ensure that every user or at least say 30% of all Termux community have read the instructions, understood them properly and implemented all steps as expected.* - That's not possible, considering the experience that only very few people actually read the posts marked as "README" or "INFO" and rest just ignores them. A such example would be https://github.com/termux/termux-packages/issues/6726.

So we have for that 2 solutions:
1. Implement a pop-up notification broadcast system in Termux app or something like a dynamic motd downloaded from our Github repo during shell startup.
2. Do all significant work ourselves, e.g. as I suggested make a script and put into `$PREFIX/etc/profile.d` and then only prompt user for something like `Do you want to configure and run ...<put description with all pros/cons here>...? (yes/no): `.

*Disclaimer: both variants are just "what could be done". None of them are planned to be implemented currently because IPFS is not a permanent solution for hosting of Termux repositories in a long term.*

**Comment by SDRausty** [comment]
> only very few people actually read the posts marked as "README" 

This is hereditary;  I am attempting to integrate the [pages](https://pages.github.com/) as a `README.md` to try to resolve this important issue that you mention.

> we have for that 2 solutions

Both sound like good solutions:
1 `motd`
2 `profile.d`

> IPFS is not a final solution for hosting of Termux repositories in a long term.

Users might want to play with IPFS as a potential short term solution.

> not a final solution 

Do you see any canditates on the horizon that might fit the category as a final solution?

**Comment by xeffyr** [comment]
@SDRausty Short term solution is what we use currently, i.e. available Termux mirrors.

Long term solution is a host dedicated to Termux project and managed by Termux team members. It can be a VPS or SaaS platform like Bintray was. Whatever it will be, members of @termux org should have right to upload packages on it. This is what is missing on mirrors where only their owner has right to upload packages and only if has access to GPG private key for signing metadata.

Please follow the full discussion to understand what is going on.

**Comment by buttaface** [comment]
> If @xeffyr doesn't want to manage the finances, who else can? buttaface can you?

Nowadays you don't really have to, as there are [transparent crowdfunding options, like this one by the Linux Foundation, where all money raised is publicly displayed and all expenses are publicly listed](https://lfx.linuxfoundation.org/tools/crowdfunding), so there is really nothing to "manage" or anybody to trust ([an example project using that software can be seen here](https://crowdfunding.lfx.linuxfoundation.org/projects/graphql)). Of course, somebody has to set that up, which I could do if the main Termux team leaders don't want to.

**Comment by Grimler91** [comment]
Thanks for the write-up @agnostic-apollo. 

Regarding 2. we have two points basically, since grimler.se is updated straight from CI (when that works..) and not mirrored from xeffyrs repos. 

If we use the funding to setup an official vps and use that as main repo then we could ask some of the big mirrors if they could mirror those repositories. My closest mirror site, [acc.umu.se](https://ftp.acc.umu.se/about/index.html) for example, indicate that they are willing to help open source projects (if we provide information, try to foresee the future, and implement some geo-based mirror selecting system). 

If we have an official vps I imagine that all members (that want it) can get user access to it, and thereby possibility to manually add debs and check the status. Preferably we would have some safe way to handle interactions with the mirror so that everything cannot be accidentally deleted..

If we ask for donations, and start using the patreon money to help the project, then it might be nice to keep a public ledger of incomes and expenses as well so that donating people know what the money is used for. @buttaface's example above looks like a good example (more serious/official looking than I had in mind).

**Comment by buttaface** [comment]
> My only concern is that account creation is a huge turn off, specially would be when donating

Yes, that is a big problem with LFX, I was just citing that as one example. [This other fundraising site doesn't require an account for one-time donations (it does for recurring donations)](https://www.oscollective.org/) and seems to raise money for a lot more OSS projects, but charges 10%. I think that would be worth it for not requiring account creation.

> People here are being "careful"

Makes sense any time money is involved, but that's also why this isn't like open source, where anybody can step up and contribute. You need someone trusted to handle money matters, which is why it probably has to be @fornwall or someone he delegates.

**Comment by agnostic-apollo** [comment]
Well, www.oscollective.org looks good too. Even F-Droid is hosted there. Great that non-account payments are accepted too credit card/paypal/bank transfer and 10% should be worth it. Their financial management compared to LFX would have to be checked as well, hopefully is fine too, specially considering there are charging for it, which is more sustainable.

But not sure if we will be able to meet their hard requirements

>If you are an open source project with at least 100 stars on GitHub and at least two contributors, you will likely be immediately approved.

>Makes sense any time money is involved... You need someone trusted to handle money matters

Yeah, that's why I said understandable, can be a big responsibility, but as discussed, if crowdfunding site is managing most of the finances, then may not be too big. 

>which is why it probably has to be @fornwall or someone he delegates.

Yeah, that would be ideal.

**Comment by fornwall** [comment]
For short-term, perhaps we just need a beefy enough VPS (or something like a Hetzner dedicated server), and hand out ssh access to those who takes on administering it? If so I can set that up.

If we're hosting packages ourself I still think sponsorship from a CDN like Cloudflare would be ideal, if possible:

1. It saves so much time and energy to get a stable solution up and running compared to setting up and administering a global mirroring system.
2. It's really convenient for users from different parts of the world (Termux is big in e.g. Indonesia and Nigeria) with stable and quick access.

Of course, if we can't get sponsorship from a CDN provider it's just out of the question (since paying a good CDN for XX TB is so expensive), but I think it's still worth exploring (I previously reached out to Cloudflare, but didn't reach the right person there to have any conversation, and there are other possible CDN sponsors).

For longer-term, I think it's important to settle if "packages via APK:s" is the way forward (to overcome https://github.com/termux/termux-app/issues/1072).

If so, it would mean that Google Play will serve a majority of the bandwidth - the bandwidth requirements (for e.g. a custom F-Droid repo for non-Play users) will go down considerably and make things much easier to manage. And we need to focus on reaching a working solution there.

**Comment by a1batross** [comment]
>or something like a Hetzner dedicated server

My mirror runs on Hetzner, the bandwidth is good for me: server uploaded around 10TB data last month. It's a cheapest Ryzen server, located in Finland.

**Comment by selurvedu** [comment]
Have you considered FOSSHost? I think Termux may be eligible for their services.

https://fosshost.org/
https://docs.fosshost.org/en/home/application
https://docs.fosshost.org/en/home/mirrors-as-a-service

**Comment by xeffyr** [comment]
@selurvedu We don't need mirror-as-a-service. As I wrote here - we need either VPS or SaaS platform for hosting packages (similar to Bintray).

Mirrors are good, but please understand that they are mostly back-ups. We don't have primary server for further replication of repositories on mirrors and have to do that on our own (https://grimler.se/termux/ and IPFS nodes) and that has result of 2 branches of Termux repositories.

According to application form, they seem to not provide x86 servers (though it is possible to request AArch64 VPS as requested type of service):
> Thank you for your interest in Fosshost. We are currently not accepting any new applications for x86 Virtual Machines.

@kcubeterm said that he works on solution. It is currently available at https://dl.kcubeterm.me/termux-packages-24/ though it works in a mirror mode currently and I don't know how long term it would be.

**Comment by iBug** [comment]
@xeffyr Why is this issue closed? Do I assume you already have a long-term solution (IPFS)? It seems to me like a antipattern to close a ticket before the real issue has been resolved.

**Comment by xeffyr** [comment]
@iBug This issue has been about Bintray support finish and finding a solution to prevent service interruption. This has been solved for few months, since Feb 2021.

I will open a new issue regarding hosting options.

**Comment by selurvedu** [comment]
@xeffyr I see. You could use an AArch64 host for uploading packages _and_ a mirror for distributing them. Also it is not directly stated how their mirroring works, maybe it's simply accessible via (S)FTP.

I'm sorry that I didn't read this whole thread, so I kindly ask you to clarify what are the exact requirements for hosting a repository, i.e., do you need any specific software, or do you simply need a webhook for uploading files via HTTP from a CI? Is some kind of artifact version management needed? Do you want a feature to auto-update the repo lists when packages get added/updated/deleted, like Open Build Service does? etc.

**Comment by SDRausty** [comment]
> I'm sorry that I didn't read this whole thread, so I kindly ask you to clarify what are the exact requirements for hosting a repository, 

Can a summary of this issue be published at https://github.com/termux/termux-packages/wiki so we have a firmer grasp of this topic please?

**Comment by agnostic-apollo** [comment]
A new issue #6846 has already been opened with the details.

**Comment by MC45892** [comment]
Termux could use GitHub pages

**Comment by insign** [comment]
> Termux could use GitHub pages

GH has many kinds of limits since it is not for that.
